File: 0023 合并K个升序链表/main.py
class Solution:
    
    def mergeKLists(self, lists: List[ListNode]) -> ListNode:
        if len(lists)==0:
            return None
        orderedL = lists[0]
        
        def merge( l,r): 
            
            mNode=ListNode(0)
            root=mNode
            while l and r:
                if l.val < r.val:
                    mNode.next=ListNode(l.val)
                    l=l.next
                    mNode=mNode.next
                else:
                    mNode.next=ListNode(r.val)
                    r=r.next
                    mNode=mNode.next
            if l:
                mNode.next = l
            else:
                mNode.next = r
            return root.next
        
        while len(lists) > 1:
            mergedL=[]
            if len(lists)%2==1:
                mergedL.append(lists.pop())
                print(lists)
            for idx in range(len(lists)//2):
                
                mergedL.append(merge( lists[2 * idx] ,lists[2 * idx+1]))
            lists=mergedL
        return lists[0]

File: 0023 合并K个升序链表/main.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

struct comp{
    bool operator() (ListNode * const a,ListNode * const b){
        return a->val>b->val;
    }
};

class Solution {
public:
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        ListNode head;
        ListNode * node=&head;
        priority_queue<ListNode *,vector<ListNode *> ,comp > h;
        for(int i = 0;i<lists.size();i++){
            if(lists[i]!=nullptr)
                h.push(lists[i]);
        }
        while( h.size() >0 ){
            node->next=h.top();
            node=node->next;
            h.pop();
            if( node->next != nullptr)
                h.push(node->next);
        }
        return head.next;
    }
};